### PR TITLE
Restore docs coverage test via sphinx

### LINKED ISSUE #37590: Rewrite documentation coverage tests
## 📚 Documentation

PR gh-37419 disabled the tests in `tests/test_doc_coverage.py` as explained in [this comment](https://github.com/pytorch/pytorch/pull/37419#issuecomment-621431059). Rather than delay that PR, this issue was opened to track refactoring those tests.

### PR DESCRIPTION
**Pull request by mattip**
Fixes gh-37590

Adds an extra `make coverage` to document building, which uses the built-in facility in sphinx to check docstring coverage. Also fixes a failure to import `torch/jit/supported_ops.py` which broke the [Torchscript Builtins](https://pytorch.org/docs/stable/jit_builtin_functions.html) page.

This also adds the required `SPHINXOPTS` to turn warnings into error, but this is commented out. Note that since documentation of `torchvision` is merged in here, failures there would cause failures here if this is made active. Some thought might be needed about pinning the torchvision version merged into documentation.

The first commit should fail, since the "ScriptModule" class is commented out. I did that in order to check that a CI failure is properly reported.

### CODE REVIEW

**Comment by mattip** [comment]
`ci/circleci: pytorch_python_doc_push` failed as expected on b9e9521, passes on 469bd98. Flake8 failure is unrelated, is there a way to rerun it?

**Comment by mattip** [comment]
Two failing CI runs  are unrelated:
- `pytorch_linux_xenial_py3_6_gcc5_4_test` does not like mypy,
- `py3.6-clang7-rocmdeb-ubuntu16.04-test2` failed with
  ```
  test_broadcast_coalesced (__main__.TestCuda) ... \
    Memory exception on virtual address   0xbeefdeadb000, node id 5 : Page not present
  06:00:06 Address does not belong to a known buffer
  06:00:06 Memory access fault by GPU node-5 (Agent handle: 0x31e7960) \
    on address   0xbeefdeadb000. Reason: Page not present or supervisor privilege.
  ```

**Comment by mattip** [comment]
rebased to clear merge conflicts

**Comment by rgommers** [comment]
The one CI failure is an NCCL reduce issue, looks unrelated.

> Flake8 failure is unrelated, is there a way to rerun it?

For re-running all tests see https://github.com/pytorch/pytorch/wiki/Bot-commands, for just one you may have a button in CircleCI if you log in there with your GitHub username.

File: docs/source/conf.py [hunk 1]
```diff
@@ -78,6 +78,56 @@
 if RELEASE:
     templates_path = ['_templates-stable'] + templates_path
```
Comment (rgommers) [inline docs/source/conf.py:80]: Probably good to add a comment here with some details on why things are ignored here.

File: .circleci/scripts/python_doc_push_script.sh [hunk 2]
```diff
@@ -71,7 +71,19 @@ cp -a ../vision/docs/source source/torchvision
 # Build the docs
 pip -q install -r requirements.txt || true
 if [ "$is_master_doc" = true ]; then
+  # export SPHINXOPTS="-WT --keep-going"
```
Comment (rgommers) [inline .circleci/scripts/python_doc_push_script.sh:74]: Remove, or uncomment?

Comment (rgommers) [reply .circleci/scripts/python_doc_push_script.sh:74]: This block, especially `lines-$(wc...`, is a little hard to read, a comment here would also be helpful.

**Comment by ezyang** [comment]
I am happy to land this when @rgommers thinks it's ready!

File: docs/source/conf.py (hunk 1 above)
Comment (mattip) [reply docs/source/conf.py:80]: adding a TODO comment

File: .circleci/scripts/python_doc_push_script.sh (hunk 2 above)
Comment (mattip) [reply .circleci/scripts/python_doc_push_script.sh:75]: Adding a comment block and a TODO on the `SPHINXOPTS="-WT --keep-going"`, with a reference to gh-38011

**Review by mattip — COMMENTED** [review]
Fixed in c4c6d8c

**Review by rgommers — APPROVED** [review]
Comment is clear and coverage is run correctly:
```
May 19 11:42:38 Testing of coverage in the sources finished, look at the results in build/coverage/python.txt.
May 19 11:42:38 +++ wc -l build/coverage/python.txt
May 19 11:42:38 +++ cut -f1 '-d '
May 19 11:42:38 ++ lines=2
May 19 11:42:38 ~/workspace ~/workspace/pytorch.github.io ~/workspace
May 19 11:42:38 ++ undocumented=0
May 19 11:42:38 ++ '[' 0 -lt 0 ']'
May 19 11:42:38 ++ '[' 0 -gt 0 ']'
```

Over time hopefully the coverage can be extended so there's less in `coverage_ignore_modules` - that may be worth opening a new issue for @mattip?

This PR LGTM.

File: docs/source/conf.py [hunk 3]
```diff
@@ -78,6 +78,58 @@
...
+    "torch.hub",
+    "torch.jit.unsupported_tensor_ops",
+    "torch.onnx",
+    "torch.nn.quantized.functional",
+    "torchvision",
+]
+
+coverage_ignore_functions = [
+    # torch.jit
+    "annotate",
```
Comment (vadimkantorov) [inline docs/source/conf.py:97]: "annotate" somehow is mentioned twice

Comment (mattip) [reply docs/source/conf.py:97]: fixed in b8ab5ce380

**Pull request closed without merging**